File: 014_CommonPrefix.py
#Method from others : https://discuss.leetcode.com/topic/6308/simple-python-solution/2
#13.88%
class Solution:
    # @return a string
    def longestCommonPrefix(self, strs):
        if not strs:
            return ""
            
        for i, letter_group in enumerate(zip(*strs)):
            if len(set(letter_group)) > 1:
                return strs[0][:i]
        else:
            return min(strs)

#Another method from others: https://discuss.leetcode.com/topic/6308/simple-python-solution/2
#57.73%
class Solution:
    # @return a string
    def longestCommonPrefix(self, strs):
        def lcp(s, t):
            if len(s)>len(t):
                s, t = t, s
            for i in range(len(s)):
                if s[i]!=t[i]:
                    return s[:i]
            return s
        return reduce(lcp,strs) if strs else ""

#26.99% (my method. a shame to python. ^_^)
class Solution(object):
    def longestCommonPrefix(self, strs):
        """
        :type strs: List[str]
        :rtype: str
        """
        if len(strs) == 0:
            return ""
        if len(strs) == 1:
            return strs[0]
        
        flag = True
		pos = 0
        for charFirst in strs[0]:
            for str in strs[1:]:
                if len(str) <= pos or str[pos] != charFirst:
                    flag = False
                    break
            if not flag:
                break
            pos = pos + 1
        
        return strs[0][:pos]

File: 014_CommonPrefix.cpp
class Solution {
public:
    string longestCommonPrefix(vector<string>& strs) {
        //Boundary conditions check
		int size = strs.size();
		if(0 == size)
			return "";
		if(1 == size)
			return strs[0];
		
		bool flag = true;
		int sizeFirst = strs[0].size();
		int pos = 0;
		for(pos = 0; pos < sizeFirst; pos++)
		{
			char charFirst = strs[0][pos];
			for(int indexStrs = 1; indexStrs < size; indexStrs++)
				if(pos >= strs[indexStrs].size() || strs[indexStrs][pos] != charFirst)
				{
					flag = false;
					break;
				}
			if(!flag)
				break;
		}
		
		return strs[0].substr(0, pos);
    }
};

//17.11%
class Solution001 {
public:
    string longestCommonPrefix(vector<string>& strs) {
        //Boundary conditions check
		int size = strs.size();
		if(0 == size)
			return "";
		if(1 == size)
			return strs[0];
		
		bool flag = true;
		int sizeFirst = strs[0].size();
		int pos = 0;
		for(pos = 0; pos < sizeFirst; pos++)
		{
			char charFirst = strs[0][pos];
			for(int indexStrs = 1; indexStrs < size; indexStrs++)
				if(pos >= strs[indexStrs].size() || strs[indexStrs][pos] != charFirst)
				{
					flag = false;
					break;
				}
			if(!flag)
				break;
		}
		
		return strs[0].substr(0, pos);
    }
};